File: src/main/java/celtric/katas/game_of_life_2017_05/Game.java
package celtric.katas.game_of_life_2017_05;

import java.util.HashSet;
import java.util.Set;
import java.util.stream.Collectors;

public final class Game {

    private final Set<Coordinate> aliveCells;

    public Game() {
        this.aliveCells = new HashSet<>();
    }

    private Game(Set<Coordinate> aliveCells) {
        this.aliveCells = aliveCells;
    }

    public void addCell(int x, int y) {
        aliveCells.add(new Coordinate(x, y));
    }

    public void removeCell(int x, int y) {
        aliveCells.remove(new Coordinate(x, y));
    }

    public boolean isAlive(int x, int y) {
        return aliveCells.contains(new Coordinate(x, y));
    }

    public Game evolve(int times) {
        return times == 1 ? evolve() : evolve(times - 1).evolve();
    }

    public Game evolve() {
        return new Game(aliveCells.stream()
                .flatMap(c -> c.neighbours().stream())
                .filter(c -> c.isNextGenerationAlive(aliveCells))
                .collect(Collectors.toSet()));
    }

    @Override
    public boolean equals(Object obj) {
        if (obj instanceof Game) {
            return aliveCells.equals(((Game) obj).aliveCells);
        } else {
            return super.equals(obj);
        }
    }
}

File: src/main/kotlin/celtric/katas/game_of_life_2017_05/Game.kt
package celtric.katas.game_of_life_2017_05

import java.util.*

data class Game (private val aliveCells: MutableSet<Coordinate> = emptySet<Coordinate>().toMutableSet()){
    fun addCell(x: Int, y: Int) {
        aliveCells.add(Coordinate(x, y))
    }

    fun removeCell(x: Int, y: Int) {
        aliveCells.remove(Coordinate(x, y))
    }

    fun isAlive(x: Int, y: Int): Boolean {
        return aliveCells.contains(Coordinate(x, y))
    }

    fun evolve(times: Int): Game {
        return if (times == 1) evolve() else evolve(times - 1).evolve()
    }

    fun evolve(): Game {
        return Game(
                aliveCells
                        .flatMap { c -> c.neighbours() }
                        .filter { c -> c.isNextGenerationAlive(aliveCells) }
                        .toMutableSet())
    }
}
